source/LiteralModel.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <mariana-trench/EventLogger.h>
+#include <mariana-trench/LiteralModel.h>
+#include <mariana-trench/Positions.h>
+
+namespace marianatrench {
+
+namespace {
+
+class LiteralModelConsistencyError {
+ public:
+  static void raise(const std::string& what) {
+    ERROR(1, "Literal Model Consistency Error: {}", what);
+    EventLogger::log_event("regex_model_consistency_error", what);
+  }
+};
+
+}
+
+static void check_taint_config_consistency(const std::string& pattern, const TaintConfig& frame) {
+  if (frame.kind() == nullptr) {
+    LiteralModelConsistencyError::raise(fmt::format(
+        "Model for regex `{}` must have a source kind.", pattern));
+  }
+}
+
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
+    check_taint_config_consistency(pattern, source);
+    all_sources.join_with(Taint{source});
+  }
+  return all_sources;
+}
+
+LiteralModel::LiteralModel() { }
+
+LiteralModel::LiteralModel(
+    std::string pattern,
+    const std::vector<TaintConfig>& sources)
+    : regex_(pattern), sources_(create_taint(regex_->pattern(), sources)) {
+}
+
+LiteralModel::LiteralModel(const LiteralModel& other)
+    : regex_(other.regex_ ? std::optional(other.regex_->pattern()) : std::nullopt), sources_(other.sources_) {
```

Comment:
I guess this is because `re2::RE2` is not copyable? That's annoying.

Reply:
Correct:
```c++
  // Not copyable.
  // RE2 objects are expensive. You should probably use std::shared_ptr<RE2>
  // instead. If you really must copy, RE2(first.pattern(), first.options())
  // effectively does so: it produces a second object that mimics the first.
  RE2(const RE2&) = delete;
```

Reply:
I think that's fine as is :)